### PR TITLE
:new: Receipt trie building

### PR DESCRIPTION
**Pull request by tongzhi1998**
- Add Receipt Trie building & storing mechanisms

### CODE REVIEW

File: include/monad/trie/rocks_writer.hpp [hunk 1]
```diff
@@ -55,6 +58,18 @@ struct RocksWriter
...
     }
+
+    void clear()
+    {
+        std::unique_ptr<rocksdb::Iterator> it(
+            db->NewIterator(rocksdb::ReadOptions{}, cf));
+        rocksdb::Status res;
+
+        for (it->SeekToFirst(); it->Valid(); it->Next()) {
+            res = db->Delete(rocksdb::WriteOptions{}, cf, it->key());
```
Comment (tongzhi1998) [inline include/monad/trie/rocks_writer.hpp]: I know this is not ideal, but using batch will can't achieve the purpose of getting rid of the entry in the DB immediately, and that creates a circular dependency between {clear(), Write() and root_hash()}.

File: include/monad/trie/key_buffer.hpp [hunk 2]
```diff
@@ -20,6 +20,13 @@ struct KeyBuffer
     size_t buf_size = 0;
     size_t prefix_size = 0;
 
+    void clear()
+    {
+        buf_size = 0;
+        prefix_size = 0;
+        memset(raw.data(), 0, sizeof(raw));
```
Comment (rgarc) [inline include/monad/trie/key_buffer.hpp]: In case you find memset error-prone, I think you could also do
```C++
raw.fill(0);
```

File: include/monad/db/in_memory_trie_db.hpp [hunk 3]
```diff
@@ -41,6 +41,22 @@ struct InMemoryTrieDB : public Db
         {
         }
 
+        void reset_cursor()
+        {
+            leaves_cursor.reset();
+            trie_cursor.reset();
+            leaves_storage.clear();
+            trie_storage.clear();
+        }
```
Comment (zander-xyz) [inline include/monad/db/in_memory_trie_db.hpp]: This is doing two separate things
- resetting the cursor (which is a noop so these calls are not necessary)
- resetting the storage

This function ought to be named more accurately

File: include/monad/db/in_memory_trie_db.hpp [hunk 4]
```diff
@@ -41,6 +41,22 @@ struct InMemoryTrieDB : public Db
...
+
+        void clear_writer()
+        {
+            trie_writer.clear();
+            leaves_writer.clear();
+        }
+
+        void clear_buffer() { trie.clear_buffer(); }
```
Comment (zander-xyz) [inline include/monad/db/in_memory_trie_db.hpp]: why are these needed?

File: include/monad/db/rocks_trie_db.hpp [hunk 5]
```diff
@@ -114,6 +120,14 @@ struct RocksTrieDB : public Db
             trie_cursor.reset();
         }
 
+        void clear_writer()
+        {
+            leaves_writer.clear();
+            trie_writer.clear();
+        }
+
+        void clear_buffer() { trie.clear_buffer(); }
```
Comment (zander-xyz) [inline include/monad/db/rocks_trie_db.hpp]: ditto question - why is this needed?

File: include/monad/db/trie_db_process_changes.hpp [hunk 6]
```diff
@@ -110,6 +116,19 @@ void trie_db_process_changes(
...
+
+    // Receipts Trie Building
+    if (!obj.receipts.empty()) {
+        for (auto i = 0u; i < obj.receipts.size(); ++i) {
+            receipt_trie_updates.emplace_back(trie::Upsert{
+                .key = trie::Nibbles{rlp::encode_unsigned(i)},
+                .value = rlp::encode_receipt(obj.receipts[i])});
+        }
+        std::ranges::sort(
+            receipt_trie_updates, std::less<>{}, trie::get_update_key);
```
Comment (zander-xyz) [inline include/monad/db/trie_db_process_changes.hpp:126]: I have not looked too in depth into how we generate receipts - is it necessary to sort (ie. does it come presorted)?

File: include/monad/state/state.hpp [hunk 7]
```diff
@@ -306,15 +307,22 @@ struct State
                code_.can_commit();
     }
 
+    void add_receipts(std::vector<Receipt> &receipts)
+    {
+        receipts_ = std::move(receipts);
+    }
```
Comment (zander-xyz) [inline include/monad/state/state.hpp]: if the intention is to transfer ownership, just take in `receipts` by rvalue reference to make it clear that this is happening (otherwise risk dangling pointer errors)

File: include/monad/execution/replay_block_db.hpp [hunk 8]
```diff
@@ -158,18 +157,10 @@ class ReplayFromBlockDb
...
-
-                auto const transactions_root = transaction_trie.root_hash();
-                auto const receipts_root = receipt_trie.root_hash();
-
                 if (!verify_root_hash(
                         block.header,
-                        transactions_root,
-                        receipts_root,
+                        NULL_ROOT, // txn root
+                        state.get_receipt_hash(),
```
Comment (zander-xyz) [inline include/monad/execution/replay_block_db.hpp:163]: I don't see us using `add_receipt` anywhere except for int he unit tests. So this `verify` will always just log a mismatch. Can we add this in?

File: include/monad/trie/in_memory_writer.hpp [hunk 9]
```diff
@@ -67,6 +67,14 @@ struct InMemoryWriter
         prefix_deletions.clear();
         deletions.clear();
     }
+
+    void clear()
+    {
+        storage.clear();
+        upserts.clear();
+        prefix_deletions.clear();
+        deletions.clear();
```
Comment (zander-xyz) [inline include/monad/trie/in_memory_writer.hpp]: This function should just be `prefix_deletions.emplace({})`, and then let `write` actually do the deletion when that gets called

File: include/monad/trie/key_buffer.hpp [hunk 10]
```diff
@@ -20,6 +20,13 @@ struct KeyBuffer
     size_t buf_size = 0;
     size_t prefix_size = 0;
 
+    void clear()
+    {
+        buf_size = 0;
+        prefix_size = 0;
+        raw.fill(0);
+    }
```
Comment (zander-xyz) [inline include/monad/trie/key_buffer.hpp]: why is this needed?

File: include/monad/trie/key_buffer.hpp [hunk 11]
```diff
@@ -31,7 +38,13 @@ struct KeyBuffer
...
     constexpr void set_prefix(byte_string_view prefix) noexcept
     {
-        MONAD_DEBUG_ASSERT(prefix.size() == sizeof(address_t));
+        MONAD_DEBUG_ASSERT(
+            prefix.size() == sizeof(address_t) || prefix.size() == 0);
+        if (prefix.size() == 0) {
+            prefix_size = 0;
+            buf_size = 0;
+            return;
+        }
```
Comment (zander-xyz) [inline include/monad/trie/key_buffer.hpp:40]: what is this?

File: include/monad/trie/rocks_cursor.hpp [hunk 12]
```diff
@@ -26,6 +26,8 @@ struct RocksCursor
     rocksdb::ReadOptions read_opts_;
     mutable KeyBuffer buf_;
 
+    bool force_reset = false;
```
Comment (zander-xyz) [inline include/monad/trie/rocks_cursor.hpp]: I don't think these are the changes that I meant, let's discuss in person.

File: include/monad/trie/trie.hpp [hunk 13]
```diff
@@ -57,6 +57,8 @@ struct Trie
     {
     }
 
+    constexpr void clear_buffer() { buf_.clear(); }
```
Comment (zander-xyz) [inline include/monad/trie/trie.hpp]: why?

File: include/monad/execution/replay_block_db.hpp (hunk 8 above)
Comment (tongzhi1998) [reply include/monad/execution/replay_block_db.hpp:163]: Discussed on Zoom, decided to punt on this until state2 comes out
